### ISSUE TITLE
Different loss profiles

### ISSUE THREAD

**Issue by alxndrTL** [zhixuan-lin/forgetting-transformer#2]
Hello,

Thank you very much for open-sourcing your work!
I'm amazed that with small experiments like 125M on 2.7B tokens we can already see the difference between architectures like Mamba and Transformer.

I have a question regarding the loss profiles in the different experiments that you've done.

For example, figure 16 in section F6 gives the following results :

![Image](https://github.com/user-attachments/assets/9fea569b-7099-483d-95cc-23aec783d5f0)

which are very good as it shows for example that FoX (Pro) keeps using the context as it goes through the sequence, well beyond the training context length.

However, this is a very different profile from Figure 2:

![Image](https://github.com/user-attachments/assets/aa9a780c-4b4d-43ad-ae8f-f5b893a4f95f)
in which FoX (Pro) is only able to use context just a bit beyond its training context length.

Is there anything fundamentally different between the two experiments other than model size and dataset size ?
How do you explain the difference between those profiles ?

Best

**Comment by zhixuan-lin** [comment]
Hi @alxndrTL,

This is a very good observation! I'm fairly confident that this is due to different model sizes and dataset size, though the learning rates are also different in these two plots.

In general, the trend I observe is:
* More training tokens lead to **worse** extrapolation (see Figure 5 in the paper).
* Larger models have **worse** extrapolation. I'm not 100% sure about this because I don't have a controlled experiment where I train models with different sizes on the **same amount of tokens**. But I'm very confident that a combination of larger models + more training tokens would get you worse extrapolation, as shown in the plots you mentioned and also Figures 19 to 21 in the paper.
* Longer training context length leads to **better** extrapolation. Again, see Figures 19 to 21.

In addition, I find learning rate could affect extrapolation. See Figure 5. But I'm not sure that a larger learning rate always gives you worse extrapolation because I don't have enough data points.

All these make sense if you think of poor extrapolation as a result of "overfitting to the training context length" (whatever that means). 
* If you train longer, naturally, there would be stronger overfitting.
* Smaller models have smaller capacity, so less likely to overfit
* Longer sequences are harder to model, so less likely to overfit

However, I'm not sure what exactly is happening when a model "overfits to the training context length". It could be related to how much the model's prediction depends on the **absolute positions** of the tokens. It is well known that models learn absolute positional information with ALiBi (see [this paper](https://arxiv.org/abs/2203.16634)), and our method can be seen as learnable ALiBi. There is also some related research for linear complexity models you can check out, e.g., the [stuffed mamba paper](https://arxiv.org/abs/2203.16634).

**Comment by alxndrTL** [comment]
Thank you for your detailed response!
Yes ok this is was very unintuitive for me, as I would have think larger models + more data = better long context modelling perf, but this is the contrary that happens in practice. Very interesting! Yes indeed there could be a phenomenom related to some kind of overfitting to the training length with FoX.

On a side note, I just wanted to point out the quality of your paper, there is a massive amount of helpful experiments and the approach is very rigourous. Good job!

**Comment by zhixuan-lin** [comment]
>  larger models + more data = better long context modelling perf

Just adding a quick comment that this is absolutely true **within the training context length**. But extrapolation is a very different regime. The model is only trained to model the training sequences so there is no guarantee what would happen beyond the training context length.

Also, I think long-context capabilities **within the training context length** are significantly more important than extrapolation performance because, in practice, you can always fine-tune on longer sequences (though it is more expensive). In fact, if I could rewrite this paper, I might have moved all extrapolation results to the Appendix because they are just distracting the readers from the performance within the training context length, which is what really matters.

**Comment by alxndrTL** [comment]
Well I kind of agree with you, although I think length extrapolation is very interesting also and yet a scientific milestone to achieve.
But also, have you tried doing length context expansion on your FoX models ? Because with such overfitting to the training context length observed on the FoX models, maybe they are harder to fine-tune on long context data, as is commonly done today ?

**Comment by zhixuan-lin** [comment]
I haven't tried it, but I might do so very soon. My guess is it should work very well given the stable extrapolation behavior. Note that Figure 2 is pretty zoomed in, so it is not as bad as it looks. The actual amount of loss degradation beyond the training context length is only around 0.07.
